### PR TITLE
Fixed heap buffer overflow in stl_update_connects_remove_1

### PR DESCRIPTION
**Pull request by andreondra**
Fixed heap buffer overflow as stated in issue #28.

### CODE REVIEW

**Comment by gladk** [comment]
Thanks for fixing this!

Do you have a POC (proof-of-concept exploit) that demonstrates it? I intend to request a CVE to track this security issue and fix it in some older Linux distributions like Debian Stretch, Buster etc.

Thanks!

**Comment by andreondra** [comment]
You are welcome!

I do not have any exploit, but a file attached in the issue #28 triggers invalid writes in Valgrind which corresponds to the ASan log in the issue. 

Here is the output of Valgrind ran on an unpatched binary:
```
==52468== Invalid read of size 4
==52468==    at 0x485B858: stl_update_connects_remove_1 (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B5EE: stl_remove_degenerate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B242: stl_remove_unconnected_facets (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x4865093: stl_repair (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B691: main (admesh.c:383)
==52468==  Address 0x4b9b9c0 is 16 bytes before a block of size 192 alloc'd
==52468==    at 0x4848A23: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==52468==    by 0x486101F: stl_allocate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x48608FD: stl_open (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B150: main (admesh.c:330)
==52468== 
==52468== Invalid read of size 4
==52468==    at 0x485B877: stl_update_connects_remove_1 (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B5EE: stl_remove_degenerate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B242: stl_remove_unconnected_facets (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x4865093: stl_repair (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B691: main (admesh.c:383)
==52468==  Address 0x4b9b9c4 is 12 bytes before a block of size 192 alloc'd
==52468==    at 0x4848A23: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==52468==    by 0x486101F: stl_allocate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x48608FD: stl_open (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B150: main (admesh.c:330)
==52468== 
==52468== Invalid read of size 4
==52468==    at 0x485B89A: stl_update_connects_remove_1 (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B5EE: stl_remove_degenerate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B242: stl_remove_unconnected_facets (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x4865093: stl_repair (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B691: main (admesh.c:383)
==52468==  Address 0x4b9b9c8 is 8 bytes before a block of size 192 alloc'd
==52468==    at 0x4848A23: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==52468==    by 0x486101F: stl_allocate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x48608FD: stl_open (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B150: main (admesh.c:330)
==52468== 
==52468== Invalid read of size 4
==52468==    at 0x485B858: stl_update_connects_remove_1 (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B605: stl_remove_degenerate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B242: stl_remove_unconnected_facets (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x4865093: stl_repair (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B691: main (admesh.c:383)
==52468==  Address 0x4b9b9c0 is 16 bytes before a block of size 192 alloc'd
==52468==    at 0x4848A23: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==52468==    by 0x486101F: stl_allocate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x48608FD: stl_open (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B150: main (admesh.c:330)
==52468== 
==52468== Invalid read of size 4
==52468==    at 0x485B877: stl_update_connects_remove_1 (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B605: stl_remove_degenerate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B242: stl_remove_unconnected_facets (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x4865093: stl_repair (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B691: main (admesh.c:383)
==52468==  Address 0x4b9b9c4 is 12 bytes before a block of size 192 alloc'd
==52468==    at 0x4848A23: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==52468==    by 0x486101F: stl_allocate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x48608FD: stl_open (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B150: main (admesh.c:330)
==52468== 
==52468== Invalid read of size 4
==52468==    at 0x485B89A: stl_update_connects_remove_1 (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B605: stl_remove_degenerate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x485B242: stl_remove_unconnected_facets (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x4865093: stl_repair (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B691: main (admesh.c:383)
==52468==  Address 0x4b9b9c8 is 8 bytes before a block of size 192 alloc'd
==52468==    at 0x4848A23: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==52468==    by 0x486101F: stl_allocate (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x48608FD: stl_open (in /home/golas/Repos/admesh/build/libadmesh.so.1.0.0)
==52468==    by 0x10B150: main (admesh.c:330)
==52468== 
```

File: src/connect.c [hunk 1]
```diff
@@ -826,7 +825,11 @@ void
 stl_update_connects_remove_1(stl_file *stl, int facet_num) {
   int j;
 
-  if (stl->error) return;
+  if (
+    stl->error ||
+    facet_num < 0
```
Comment (hroncok) [inline src/connect.c:830]: What does `facet_num < 0` indicate here? Is it an invalid input that should be indicated by setting `stl->error`?

**Review by hroncok — COMMENTED** [review]
Before:

```
1 errors in context 1 of 6:
Invalid read of size 4
   at 0x485B375: stl_update_connects_remove_1 (in .../admesh/libadmesh.so.1.0.0)
   by 0x485B0E4: stl_remove_degenerate (in .../admesh/libadmesh.so.1.0.0)
   by 0x485AD25: stl_remove_unconnected_facets (in .../admesh/libadmesh.so.1.0.0)
   by 0x4864674: stl_repair (in .../admesh/libadmesh.so.1.0.0)
   by 0x402299: main (in .../admesh/admesh)
 Address 0x4b6a9c8 is 8 bytes before a block of size 192 alloc'd
   at 0x4849464: calloc (vg_replace_malloc.c:1328)
   by 0x48607E5: stl_allocate (in .../admesh/libadmesh.so.1.0.0)
   by 0x486015B: stl_open (in .../admesh/libadmesh.so.1.0.0)
   by 0x401DDB: main (in .../admesh/admesh)


1 errors in context 2 of 6:
Invalid read of size 4
   at 0x485B352: stl_update_connects_remove_1 (in .../admesh/libadmesh.so.1.0.0)
   by 0x485B0E4: stl_remove_degenerate (in .../admesh/libadmesh.so.1.0.0)
   by 0x485AD25: stl_remove_unconnected_facets (in .../admesh/libadmesh.so.1.0.0)
   by 0x4864674: stl_repair (in .../admesh/libadmesh.so.1.0.0)
   by 0x402299: main (in .../admesh/admesh)
 Address 0x4b6a9c4 is 12 bytes before a block of size 192 alloc'd
   at 0x4849464: calloc (vg_replace_malloc.c:1328)
   by 0x48607E5: stl_allocate (in .../admesh/libadmesh.so.1.0.0)
   by 0x486015B: stl_open (in .../admesh/libadmesh.so.1.0.0)
   by 0x401DDB: main (in .../admesh/admesh)


1 errors in context 3 of 6:
Invalid read of size 4
   at 0x485B333: stl_update_connects_remove_1 (in .../admesh/libadmesh.so.1.0.0)
   by 0x485B0E4: stl_remove_degenerate (in .../admesh/libadmesh.so.1.0.0)
   by 0x485AD25: stl_remove_unconnected_facets (in .../admesh/libadmesh.so.1.0.0)
   by 0x4864674: stl_repair (in .../admesh/libadmesh.so.1.0.0)
   by 0x402299: main (in .../admesh/admesh)
 Address 0x4b6a9c0 is 16 bytes before a block of size 192 alloc'd
   at 0x4849464: calloc (vg_replace_malloc.c:1328)
   by 0x48607E5: stl_allocate (in .../admesh/libadmesh.so.1.0.0)
   by 0x486015B: stl_open (in .../admesh/libadmesh.so.1.0.0)
   by 0x401DDB: main (in .../admesh/admesh)


1 errors in context 4 of 6:
Invalid read of size 4
   at 0x485B375: stl_update_connects_remove_1 (in .../admesh/libadmesh.so.1.0.0)
   by 0x485B0CD: stl_remove_degenerate (in .../admesh/libadmesh.so.1.0.0)
   by 0x485AD25: stl_remove_unconnected_facets (in .../admesh/libadmesh.so.1.0.0)
   by 0x4864674: stl_repair (in .../admesh/libadmesh.so.1.0.0)
   by 0x402299: main (in .../admesh/admesh)
 Address 0x4b6a9c8 is 8 bytes before a block of size 192 alloc'd
   at 0x4849464: calloc (vg_replace_malloc.c:1328)
   by 0x48607E5: stl_allocate (in .../admesh/libadmesh.so.1.0.0)
   by 0x486015B: stl_open (in .../admesh/libadmesh.so.1.0.0)
   by 0x401DDB: main (in .../admesh/admesh)


1 errors in context 5 of 6:
Invalid read of size 4
   at 0x485B352: stl_update_connects_remove_1 (in .../admesh/libadmesh.so.1.0.0)
   by 0x485B0CD: stl_remove_degenerate (in .../admesh/libadmesh.so.1.0.0)
   by 0x485AD25: stl_remove_unconnected_facets (in .../admesh/libadmesh.so.1.0.0)
   by 0x4864674: stl_repair (in .../admesh/libadmesh.so.1.0.0)
   by 0x402299: main (in .../admesh/admesh)
 Address 0x4b6a9c4 is 12 bytes before a block of size 192 alloc'd
   at 0x4849464: calloc (vg_replace_malloc.c:1328)
   by 0x48607E5: stl_allocate (in .../admesh/libadmesh.so.1.0.0)
   by 0x486015B: stl_open (in .../admesh/libadmesh.so.1.0.0)
   by 0x401DDB: main (in .../admesh/admesh)


1 errors in context 6 of 6:
Invalid read of size 4
   at 0x485B333: stl_update_connects_remove_1 (in .../admesh/libadmesh.so.1.0.0)
   by 0x485B0CD: stl_remove_degenerate (in .../admesh/libadmesh.so.1.0.0)
   by 0x485AD25: stl_remove_unconnected_facets (in .../admesh/libadmesh.so.1.0.0)
   by 0x4864674: stl_repair (in .../admesh/libadmesh.so.1.0.0)
   by 0x402299: main (in .../admesh/admesh)
 Address 0x4b6a9c0 is 16 bytes before a block of size 192 alloc'd
   at 0x4849464: calloc (vg_replace_malloc.c:1328)
   by 0x48607E5: stl_allocate (in .../admesh/libadmesh.so.1.0.0)
   by 0x486015B: stl_open (in .../admesh/libadmesh.so.1.0.0)
   by 0x401DDB: main (in .../admesh/admesh)

ERROR SUMMARY: 6 errors from 6 contexts (suppressed: 0 from 0)
```

Now:

```
ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

So this indeed fixes the problem, thanks!

One inline question.

**Comment by hroncok** [comment]
@gladk If you want to request a CVE and you know how to do it, please do. I can however live without it.

File: src/connect.c (hunk 1 above)
Comment (andreondra) [reply src/connect.c:830]: It is only an argument value check which does not have to indicate any syntactical error. The file provided in the issue is actually syntactically correct but does not make much sense (it is a line). So this fix just makes sure we don't do any invalid 
(nonsense) operations and just skip this particular procedure.

**Pull request merged**